Program.cs

```diff
@@ -3,17 +3,95 @@
    internal class Program {
+      static void Main () { }
+   }
+
+   /// <summary>Double Ended Queue class</summary>
+   public class TDeque<T> {
+      #region Property ----------------------------------------------
+      /// <summary>Gets the count of the TDeque</summary>
+      public int Count => mCount;
+
+      /// <summary>Gets the size of the TDeque</summary>
+      public int Capacity => mDeque.Length;
+
+      /// <summary>Checks whether the TDeque is empty</summary>
+      public bool IsEmpty => mCount == 0;
+      #endregion
+
       #region Method ------------------------------------------------
-      static void Main (string[] args) {
-         Console.WriteLine ("Hello, World!");
+      /// <summary>Adds the element at the rear end of the TDeque</summary>
+      /// <param name="item">Element to be added</param>
+      public void EnqueueRear (T item) {
+         if (Count == Capacity) Resize ();
+         if (IsEmpty) AddFirstElement (item);
+         else { // Setting tail pointer to free space in the start if it is at the last index.
+            mRear = (mRear == Capacity - 1) ? mRear % (Capacity - 1) : mRear + 1;
+            mDeque[mRear] = item;
+            mCount++;
+         }
+      }
+
+      /// <summary>Adds the element at the front end of the TDeque</summary>
+      /// <param name="item">Element to be added</param>
+      public void EnqueueFront (T item) {
+         if (Count == Capacity) Resize ();
+         if (IsEmpty) AddFirstElement (item);
+         else { // Setting head pointer to free space in the last if it is at the first index.
+            mFront = (mFront == 0) ? Capacity - 1 : mFront - 1;
+            mDeque[mFront] = item;
+            mCount++;
+         }
+      }
+
+      /// <summary>Removes an element from the rear end of the TDeque</summary>
+      public T DequeueRear () {
+         T dequeItem;
+         if (IsEmpty) throw new InvalidOperationException ();
+         (dequeItem, mDeque[mRear]) = (mDeque[mRear], default);
+         mCount--; // Setting back tail pointer after dequeue according to previous positions. 
+         mRear = (mRear == 0) ? Capacity - 1 : mRear - 1;
+         return dequeItem;
       }
+
+      /// <summary>Removes an element from the front end of the TDeque</summary>
+      public T DequeueFront () {
+         T dequeItem;
+         if (IsEmpty) throw new InvalidOperationException ();
+         (dequeItem, mDeque[mFront]) = (mDeque[mFront], default);
+         mCount--; // Setting back head pointer after dequeue according to previous positions. 
+         mFront = (mFront == Capacity - 1) ? 0 : mFront + 1;
+         return dequeItem;
+      }
+
+      /// <summary>Resizes the array by increasing the capacity by 2 times 
+      /// when the array is full</summary>
+      void Resize () {
+         if (mRear < mFront) // Rearranging the elements in linear way
+            mDeque = mDeque.TakeLast (Capacity - mFront).Concat (mDeque.Take (mRear + 1)).ToArray ();

```

Comment:
This should be implemented with circular buffer. No need to arrange like this and set mRear to half of the capacity. Just head should be at zero and rear as the count.